sw/host/opentitanlib/src/rom_patch/patch.rs

```diff
@@ -0,0 +1,738 @@
+    }
+}
+
+impl fmt::Display for RomPatchMatchDescriptor {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "\tMatch address {:#04x}", self.match_addr())?;
+        writeln!(f, "\tRemap address {:#04x}", self.remap_addr())?;
+        write!(
+            f,
+            "\tPatch size {:?} Locked {:?} ",
+            self.size(),
+            self.locked()
+        )
+    }
+}
+
+#[derive(Default, Clone, Copy, Debug)]
+#[allow(dead_code)]
+#[repr(packed)]
+struct RomPatchHeader {
+    meta: u8,
+    size: u16,
+    revision: u8,
+}
+
+impl Read for RomPatchHeader {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let raw_header = self as *const _ as *const u8;
+        let header_size = mem::size_of::<RomPatchHeader>();
+
+        if buf.len() < header_size {
+            return Err(std::io::Error::from(std::io::ErrorKind::InvalidInput));
+        }
+
+        unsafe {
+            ptr::copy_nonoverlapping(raw_header, buf.as_mut_ptr(), header_size);
+        }
+
+        Ok(header_size)
+    }
+}
+
+impl fmt::Display for RomPatchHeader {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let unaligned = std::ptr::addr_of!(self.size);
+        let size = unsafe { std::ptr::read_unaligned(unaligned) };
+        write!(
+            f,
+            "\tMeta {:#x} Revision {:?} Size {:?} DWORDs",
+            self.meta, self.revision, size
+        )
+    }
+}
+
+struct RomPatchSubRoutine {
+    base_addr: u32,
+    offset: usize,
+    code: Vec<u32>,
+}
+
+impl RomPatchSubRoutine {
+    fn from_hjson(routine: &RomPatchSubRoutineHjson, patch_base_addr: u32) -> Result<Self> {
+        let base_addr = u32::from_str(&routine.base_addr)?;
+        if base_addr < patch_base_addr {
+            bail!(
+                "Invalid subroutine base addr {:#x} < {:#x} ",
+                base_addr,
+                patch_base_addr
+            );
+        }
+
+        let mut code = vec![];
+        for insn in &routine.code {
+            let insn_hex = u32::from_str(insn)?;
+            code.push(insn_hex);
+        }
+
+        let offset = PATCH_CODE_SECTION_OFFSET + (base_addr - patch_base_addr) as usize;
+
+        Ok(Self {
+            base_addr,
+            offset,
+            code,
+        })
+    }
+
+    fn len(&self) -> usize {
+        self.code.len() * 4
+    }
+
+    fn offset(&self) -> usize {
+        self.offset
+    }
+
+    fn range(&self) -> Range<u32> {
+        Range {
+            start: self.base_addr,
+            end: self.base_addr + (self.code.len() as u32 * 4),
+        }
+    }
+}
+
+impl fmt::Display for RomPatchSubRoutine {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let range = self.range();
+        write!(
+            f,
+            "\tBase address {:#04x} Range [{:#04x}:{:#04x}]",
+            self.base_addr, range.start, range.end
+        )
+    }
+}
+
+impl Read for RomPatchSubRoutine {
```

Comment:
This can also be a `zerocopy::AsBytes` call I think

Reply:
Indeed. We no longer use the `Read` trait.